pkg/skaffold/build/scheduler.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package build
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/tag"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/event"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+)
+
+type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string) (string, error)
+
+type scheduler struct {
+	artifacts       []*latest.Artifact
+	nodes           []buildNode // size len(artifacts)
+	artifactBuilder ArtifactBuilder
+	logger          logAggregator
+	results         builtArtifacts
+	sem             countingSemaphore
+}
+
+func newScheduler(artifacts []*latest.Artifact, artifactBuilder ArtifactBuilder, concurrency int) *scheduler {
+	s := scheduler{
+		artifacts:       artifacts,
+		nodes:           createNodes(artifacts),
+		artifactBuilder: artifactBuilder,
+		logger:          newLogAggregator(len(artifacts)),
+		results:         newArtifactsStore(),
+		sem:             newCountingSemaphore(concurrency),
+	}
+	return &s
+}
+
+func (s *scheduler) run(ctx context.Context, out io.Writer, tags tag.ImageTags) ([]Artifact, error) {
+	g, gCtx := errgroup.WithContext(ctx)
+
+	for i := range s.artifacts {
+		i := i
+
+		// Create a goroutine for each element in dag. Each goroutine waits on its dependencies to finish building.
+		// Because our artifacts form a DAG, at least one of the goroutines should be able to start building.
+		// Wrap in an error group so that all other builds are cancelled as soon as any one fails.
+		g.Go(func() error {
+			return s.build(gCtx, tags, i)
+		})
```

Comment:
instead of passing around tags and then performing the checks later, would it make sense to change this build to just pass the artifact and node. 
```
func (s *scheduler) build(ctx context.Context, actifact, node, tag tag.ImageTag) error {

}
```
the check for `tag` exists can be moved from `performBuild` to this go subroutine.
```
for i := range s.artifacts {
   a := s.artifacts[i]
   n := nodes[i]
   g.Go(func() error {
     tag, present := tags[a.ImageName]
     if !present {
       return fmt.Errorf(...)
     }
     return s.build(gCtx, a, n, tag)
    })

}
```

Reply:
I'd like to keep it this way for 2 reasons:
- The build failure events and logs are already setup if the error is raised in `performBuild` and I don't want to repeat them elsewhere
- I've tried to reuse existing code functions as much as possible and would like to avoid changes there.

Reply:
sounds good.